sql/src/main/java/io/crate/analyze/SQLPrinter.java

```diff
@@ -265,63 +291,107 @@ private void addFrom(StringBuilder sb, QueriedRelation relation) {
+            WhereClause finalWhereClause = WhereClause.MATCH_ALL;
             // first resolve implicit cross joins, e.g. select * from t1, t2
             Iterator<AnalyzedRelation> sourcesIt = mss.sources().values().iterator();
             int numberOfCrossJoins = mss.sources().size() - mss.joinPairs().size() - 1;
             int i = 0;
             while (i < numberOfCrossJoins) {
                 if (i == 0) {
-                    printRelationName(sourcesIt.next(), sb);
+                    Symbol whereQuery = printRelationName(sourcesIt.next(), sb).query();
+                    if (whereQuery != null) {
+                        finalWhereClause = finalWhereClause.add(whereQuery);
+                    }
                 }
                 sb.append(" CROSS JOIN ");
-                printRelationName(sourcesIt.next(), sb);
+                Symbol whereSymbol = printRelationName(sourcesIt.next(), sb).query();
+                if (whereSymbol != null) {
+                    finalWhereClause = finalWhereClause.add(whereSymbol);
+                }
                 i++;
             }
             // now print explicit join pairs
             if (mss.joinPairs().isEmpty()) {
-                return;
+                return finalWhereClause;
             }
             if (i == 0) {
                 JoinPair firstPair = mss.joinPairs().iterator().next();
                 AnalyzedRelation leftRelation = mss.sources().get(firstPair.left());
-                printRelationName(leftRelation, sb);
+                Symbol whereSymbol = printRelationName(leftRelation, sb).query();
+                if (whereSymbol != null) {
+                    finalWhereClause = finalWhereClause.add(whereSymbol);
+                }
             }
             for (JoinPair currentJoinPair : mss.joinPairs()) {
                 sb.append(" ");
                 sb.append(currentJoinPair.joinType());
                 sb.append(" JOIN ");
                 AnalyzedRelation rightRelation = mss.sources().get(currentJoinPair.right());
-                printRelationName(rightRelation, sb);
-                if (currentJoinPair.joinType() == JoinType.CROSS) {
-                    continue;
+                Symbol whereSymbol = printRelationName(rightRelation, sb).query();
+                if (whereSymbol != null) {
+                    finalWhereClause = finalWhereClause.add(whereSymbol);
+                }
+                if (currentJoinPair.joinType() != JoinType.CROSS) {
+                    sb.append(" ON ");
+                    sb.append(printSymbol(currentJoinPair.condition()));
                 }
-                sb.append(" ON ");
-                sb.append(printSymbol(currentJoinPair.condition()));
             }
-
+            return finalWhereClause;
         }
 
-        private static void printRelationName(AnalyzedRelation relation, StringBuilder sb) {
+        private WhereClause printRelationName(AnalyzedRelation relation, StringBuilder sb) {
```

Comment:
That `printRelationName` returns a `WhereClause` is rather strange. Can we change the method name or otherwise re-structure this?